### ISSUE TITLE
Support yubikey challenge response

### ISSUE THREAD

**Issue by piratenpanda** [Kunzisoft/KeePassDX#8]
Would be nice if the app would support challenge response via the yubichallenge app

**Comment by J-Jamet** [comment]
This is not a priority because I have to solve bugs and put more basic features but why not in the future.

**Comment by J-Jamet** [comment]
There is an evolution in the settings with the version **2.5.0.0beta3** but the management of the kdbx V4 must be better managed and requires more time.

**Comment by rugk** [comment]
For reference KeePassXC implements this on the desktop and if you want code for Android OpenKeyChain supports YubiKey (in a different use case for PGP keys, but well… it's something.

**Comment by rugk** [comment]
Hmm… don't know whether depending on another app is so good. At least it would be nice to not only support this one vendor, but well… future plans anyway. But thanks FYI.

**Comment by piratenpanda** [comment]
If there is an implementation like in openkeychain even better, sure

**Comment by ovz93br43v7** [comment]
This feature would be great and please compatible to the KeepassXC implementation.

**Comment by piratenpanda** [comment]
might as well use ykdroid instead of the yubichallenge app as keepass2android does for a while now. Just to update on my first post

**Comment by gappleto97** [comment]
Is this still a planned feature?

**Comment by J-Jamet** [comment]
Yes, I'm just going to release the final 2.5 version before starting big jobs like this issue. I must also look at the other existing physical keys (open source) and study the functioning of KeePassXC.

**Comment by dimmuboy** [comment]
Is there any forecast which version could be have finally yubikey support?

**Comment by J-Jamet** [comment]
I have not yet looked at the technical operation of the yubikey for lack of time because of new bugs and more important features to implement, but if a technician is motivated to make a pull request, it will be very nice.

**Comment by J-Jamet** [comment]
I bought a Yubikey4 and an Onlykey to be able to test, I still have to study how the challenge response works with these keys and think about the architecture.

**Comment by hughobrien** [comment]
Hi there! I'm buying KeePassDX Pro now to show my support for this issue!
I was using passwdsafe which implements it via NFC.

Thanks for your work.

**Comment by invalid-error** [comment]
> Hi there! I'm buying KeePassDX Pro now to show my support for this issue!
> I was using passwdsafe which implements it via NFC.
> 
> Thanks for your work.



> Hi there! I'm buying KeePassDX Pro now to show my support for this issue!
> I was using passwdsafe which implements it via NFC.
> 
> Thanks for your work.

I would like do the same but I am using f-droid (no Pro available). Any update on Yubikey support (like KeePassXC Desktop). I would also like to do a donation for this new feature :1st_place_medal: .

**Comment by invalid-error** [comment]
All my BAT to you @J-Jamet ;)
... you should add this to your Cryptocurrency donation section.

**Comment by schmitmd** [comment]
Bought KeePassDX Pro and would love to see this feature.

**Comment by mariodsantana** [comment]
I use F-Droid so can't buy pro, but I just donated E10 because thank you and to support this feature!  I have to use different software until this feature makes it into KeePassDX.  :(

**Comment by J-Jamet** [comment]
Thanks for your support, the feature is planned for version 3.3.0. 
I am aware that many users want this feature, I just need to do things in order because otherwise there will be bugs in the app. I prefer to do things properly by studying the technical functioning correctly and don't add any security holes.
https://github.com/Kunzisoft/KeePassDX/projects/43
So for the moment I'm concentrating on version 3.0.0 which restructures the application for a better code architecture (so also indirectly prepares this feature).
I would also like to be able to use my keys with KeePassDX, it just takes time and working with specific hardware is different than working with only the software side.

**Comment by mariodsantana** [comment]
Understood.  I'll be waiting patiently.  :)

**Comment by invalid-error** [comment]
For me the same, I just don't use the app now since I need to sync. my database and I do not sync. it without a challenge response. I really looking forward to this implementation. I will also wait patiently.

**Comment by wts42** [comment]
Got me the pro version now and in happy anticipation of the yubikey feature. (100 euros, 1 yubikey, 1 spare yubikey and KeepassDX Pro) Money good invested. :blush:

**Comment by J-Jamet** [comment]
As I said in my previous comment, things are done in order to have as few bugs as possible. Of course, the integration of physical keys is still planned. The project section allows you to see all the work to be done and planned for future versions.
Thanks for your patience.

**Comment by phnx47** [comment]
I bought Pro version. But now, I started using Yubikey and switched to Keepass2android, because I can't unlock my db. I see that is planned, but I can't just wait

**Comment by uduerholz** [comment]
I hacked together a working prototype. "Working" means: I can open a database with my Yubikey via NFC in combination with the ykdroid app (needs to be installed first). The database was created with KeePassXC on Ubuntu.

@J-Jamet: If you like, I can clean the code, do more tests and create a pull request, but I would need to know, 
which direction to take:
Is it okay to depend on the ykdroid app? Or maybe you want to simply integrate the code (it has the same license, 
so shouldn't be a problem in that respect).

**Comment by J-Jamet** [comment]
@uduerholz Awesome!
No problem for the ykdroid dependency, that looks very good. It's also an open source app so it will be easier for maintenance.

**Comment by uduerholz** [comment]
I just stumbled upon a small usability problem. The current implementation creates a new random transform seed (part of the database header) each time the database is saved. When the master key is derived, the algorithm sends this transform seed as challenge to the Yubikey and uses the response. In effect this means that each time the database is saved, the app needs access to the Yubikey. Via USB this is not a problem, but via NFC it could be annoying.

I see the following solutions:
1. When the database is protected with a Yubikey, never modify the transform seed (makes it a bit less secure).
2. Require of the user to present the Yubikey each time the database is saved.
3. When the Yubikey is used the first time, create some random seeds in advance and send them to the Yubikey. Store the responses and use them later when the database is saved. (Does not work with the current ykdroid implementation though.)

**Comment by J-Jamet** [comment]
I imagine that you want to implement the KeeChallenge Response of KeePassXC.

Solution 2 is the most relevant. KeePassXC does require the key to be present for each save in the database, so users will not be lost.

It would be nice to be able to easily select the credentials to unlock the database. I will think about a correct UI, because I would like to implement other algorithms in the future (hmac-sha1 kpxc, hmac-sha1 kp2, oath hotp, hmac-secret fido2).

**Comment by uduerholz** [comment]
Yes, one main requirement is to be compatible with KeePassXC. When you have a suggestion for the UI, please post it here.

**Comment by J-Jamet** [comment]
For the UI, I was thinking of simply adding a field with a switch button below the key file field. If compatible drivers or applications are detected (so with an intent test), (The field becomes visible, otherwise is hidden. Maybe all the time displayed at first, this is to not lose new users as it is an advanced feature but maybe the best is to hide the key-file and challenge-response field in a section to be extended.)

It would be composed of a dropdown list to select the type of challenge protocol used. (At first, can just be a label if there is only the hmac-sha1 kpxc).

For the recovery mode, a small icon to the right of the field. When clicked, would display an edit text field below the challenge-response field.

If the recovey mode text field is [hidden] or [displayed and its content is empty], we go through ykdroid otherwise we test the number of characters requested (we display an error if necessary), and we enter directly the recovery key if the size is good.

What do you think about it?

**Comment by uduerholz** [comment]
Sounds good. What do you mean by "recovery mode"?

**Comment by J-Jamet** [comment]
By recovery mode, I mean a way to open the database without a physical key. When configuring a yubikey in challenge-response, a key is generated which is normally saved on paper in case of loss. It can be used in the recovery mode to open the file without a physical key.

**Comment by piratenpanda** [comment]
Manually entering the challenge response most likely. At least that's how it's handled elsewhere

**Comment by uduerholz** [comment]
Okay, I see. I thought you talk about something that already exists. ;-)

**Comment by J-Jamet** [comment]
The recovery mode does not need to be implemented at the beginning. The idea is that you can recreate a physical key with the same behavior from the backup, so it's just a help to not have to buy new hardware in case of loss.

**Comment by uduerholz** [comment]
I am about to implement solution 2) which is compatible to KeePassXC.
Each time the database is (automatically) saved, the activity from ykdroid is started. For NFC this means the user is asked to tap the yubikey again. The yubikey responds to the challenge and ykdroid passes the response back to the app. At that point a new master key has to be derived, which means the password has to be stored in memory. I think that's okay, because the current master key is also stored in memory, so that doesn't make a real difference, or?

**Comment by uduerholz** [comment]
To be clear: solution 1) is also compatible with KeePassXC

**Comment by J-Jamet** [comment]
You do what you think is best. Indeed, the master key is accessible as long as the database is open, it allows you to reload the database so you can use it again. I haven't studied the security of this feature yet, I've indicated solution 2 because it seemed the most relevant at the time without thinking too much about the architecture.

**Comment by uduerholz** [comment]
Coming back to the UI: How should it look like, when creating a new database?

In my opinion it would make sense to treat it similar to the "advanced" (biometric) creation or unlocking of a database:
First the database is created "normally" with password and/or key file. Later you can add protection with the yubikey. It would not be possible to combine biometric and yubikey protection, the user would have to choose.
(For a database protected by the yubikey, a simple unlocking with the touch of a finger would not be possible anyway. As the transform seed in the header could have changed outside the app, unlocking always requires interaction with the yubikey.)

**Comment by J-Jamet** [comment]
Currently, the fingerprint is linked to the keystore, so basically and currently it means that if the fingerprint is used for a database, it means that the password of a database is hashed in the keystore.
The purpose of the Yubikey (and other keys) is to add a new way to unlock a database that can be added to or substituted for others part of master key (which KeePassXC does and allows you to use password and/or key file and/or yubikey)

If for you, the yubikey uses the same workflow as the fingerprint (which is only an advanced unlock, a kind of overlay that does not replace the usual unlock methods to stay compatible with KeePassXC), it means that the Yubikey will replace the fingerprint as advanced unlock to hash the password in the Keystore?

But in this case, a database created with KeePassXC with the Yubikey as part of the master key, cannot be opened with KeePassDX. I think the Yubikey/Onlykey should be part of the password change/creation component first to be compatible with KeePassXC databases.

In the database opening workflow, if this is a problem you can disable advanced unlock (fingerprint) if the yubikey switch view is checked. (I thought that the request of the yubikey could be done when the associated switch view is checked so it wouldn't be a problem if the user has configured his fingerprint, which he will use after authenticating his physical key. It can be tedious but if the user does not want to type his big password using his fingerprint and also use his physical key, it is possible in this case)

Then, in a second time, we can think about associating the fingerprint to any part of the master key (password and/or key file and or physical key). #427

**Comment by uduerholz** [comment]
Sorry, I didn't clearly say what I had in mind. My suggestion was not to use the Keystore in combination with the yubikey. The implementation should of course be compatible with KeePassXC.
I wanted to suggest that the user can either use the fingerprint xor the yubikey. Using the yubikey the password would need to be entered manually. If that is not what you want, I would need to check how to combine fingerprint and yubikey (without the need to enter the password). But it should be feasible to use both.

**Comment by agates** [comment]
> Sorry, I didn't clearly say what I had in mind. My suggestion was not to use the Keystore in combination with the yubikey. The implementation should of course be compatible with KeePassXC.
> I wanted to suggest that the user can either use the fingerprint xor the yubikey. Using the yubikey the password would need to be entered manually. If that is not what you want, I would need to check how to combine fingerprint and yubikey (without the need to enter the password). But it should be feasible to use both. 

Check out KeePassXC's implementation of TouchID for macOS which does something similar on the desktop.  Might give a good frame of reference.

**Comment by uduerholz** [comment]
Great, thank you. I was busy recently but hope to finish the pull request soon.

**Comment by J-Jamet** [comment]
Don't rush, I'll refactor as I go along, so it will be easier and more intuitive to integrate the functionality. Also it's fine if you work on an old version of the application, I can do the merge by hand afterwards.

**Comment by uduerholz** [comment]
@J-Jamet: I will probably not have time to finish this any time soon, sorry. If you want, you can have a look at my develop branch: https://github.com/uduerholz/KeePassDX/tree/develop
It can be used to open yubikey-protected databases in read-only mode.

**Comment by J-Jamet** [comment]
No problem, I understand it's not an easy feature and it requires a lot of design time. In any case, I'll be inspired by your work, it will greatly help me, so thank you ◕‿◕.

**Comment by J-Jamet** [comment]
I did the first integration of this feature inspired by your code (@uduerholz) and integrated it into a more modular architecture. I haven't tested all the cases yet, but I can connect to the YKdroid driver with the intent.
I've been having trouble with OTG (linked to #137) on my Android device, so I'll have to look at what's causing this.
I'm focusing for the moment on the Yubikey with SHA1 Challenge-response of KeePassXC, the KeePass plugin method requires a separate XML file, which is very restrictive to manage.

The idea is also to create a driver in the Pro application that will allow to manage the driver settings according to the key and the device. Here are the repo that can be useful for that :
Ykdroid : https://github.com/pp3345/ykDroid
Yubikey4java : https://github.com/Toporin/yubikey4java
Yubico-C : https://github.com/Yubico/yubico-c | https://developers.yubico.com/yubico-c/ | https://developers.yubico.com/yubikey-personalization/

**Comment by J-Jamet** [comment]
Well, it's OK in OTG but I think the driver sometimes goes into an unrecognized state, maybe because of a thread that doesn't respect the life cycle of an activity. I'll probably redo a cleaner code of the driver in kotlin but it works!
There is still a lot of work to do for a good implementation. I will take the opportunity to put the recovery mode when I do the unit tests but it is very promising.

**Comment by J-Jamet** [comment]
The implementation of this new factor involves several elements: 
 - The challenge-response protocol of the Yubikey is not standardized even if KeePassXC fully supports it. So KeePassDX will support it under certain conditions and the new driver will be integrated only in the Pro version.
 - I tested the OnlyKey with the same HMAC-SHA1 challenge-response protocol as Yubikey but KeePassXC returns an error on the key size at each creation. I didn't go further, maybe a bug in KeePassXC.
 - In parallel, I would like a standardized and open solution present by default that everyone can use ([FIDO2](https://github.com/Kunzisoft/KeePassDX/issues/304)). But the implementation of hmac-secret needs a lot of work and thinking about its implementation so any help is welcome on this topic.

I am currently working on integrating a HardwareKey when creating a master key. Once this feature is finished, I will study the hmac-secret operation of FIDO2 in more detail.

**Comment by J-Jamet** [comment]
It will take longer than I thought. I understand better your previous comment @uduerholz https://github.com/Kunzisoft/KeePassDX/issues/8#issuecomment-1013912643

Currently, to test, we are using Ykdroid to retrieve a response after a challenge. It works because we bypass the workflow a bit by reading the file header to get the seed and get the response before doing the actions in the service.
But this is not possible for the save because the seed is randomly generated in the KDF by a service action. So I have to find a way to do the challenge-response in the middle of a service action method.
I also need to create another application to replace Ykdroid that will provide the response by service (with binding) and not by intent because it is not possible to get an intent response from a service to allow saving a new key.
So a big architecture work.

**Comment by J-Jamet** [comment]
I think on the contrary that the seed should be randomize, otherwise there is no point in having it if it always stays the same. Afterwards, there is nothing to prevent from making it a parameter and asking the user not to randomize it at each save.
After I agree with you, it's a lot of work for just that but I prefer to respect the workflow, because otherwise it will be more complicated when it comes to managing FIDO2.

**Comment by uduerholz** [comment]
Yes, of course you need a random seed - but only once, when the database is created. What is the benefit, if you replace a random value with another random value?

**Comment by J-Jamet** [comment]
If a lazy user only uses the Yubikey, uploads it to a cloud and his database is compromised, he will change the passwords of his entries and resave his database with his Yubikey (with the same seed) thinking it is ok.

But if a hacker had previously recovered the generated key, he will be able to reuse it later to unlock the database with the new data even without having a Yubikey.
I use a plug on emulator that always send the same unlock key, it reproduces this behavior.

I agree, it's a pain in the ass to use, that's why I was thinking of a parameter to prevent randomizing the seed just temporarily if you make regular saves when you fill a database.

**Comment by J-Jamet** [comment]
Currently the functionality works well but I would like to improve the driver concept (for example : by implementing a hardware key emulator in a new app) and not be dependent on Ykdroid. The problem, currently, is that the calling intent is `"net.pp3345.ykdroid.intent.action.CHALLENGE_RESPONSE"`, which is very coupled to the ykdroid app and not generic enough. I will propose this intent "android.yubikey.intent.action.CHALLENGE_RESPONSE" to be implemented so that you have a choice between driver apps. Tell me what you think about it.

**Comment by J-Jamet** [comment]
> If KeepassDX doesn't get the response from the Yubikey, then some other app must do so

This is precisely the problem. If I (or someone else, whatever) create another application that delivers an answer from a Yubikey challenge then a choice will have to be made previously in the code about which application to use (ykDroid or another, so does not leave the choice to the user).
If we make a generic intent, it will be up to the user to make that choice because several applications will be compatible. It is simply a matter of convention.

> If you change the Intent, ykDroid cannot be used anymore.

That's why I would like to make a generic intent and I opened a [ticket](https://github.com/pp3345/ykDroid/issues/21).
If another application is created to challenge Yubikey (for example an official Yubico application), it will clearly not use an intent that contains a package name from another application that does the same thing. ;)

Also, I wonder if we should not make a generic intent for the challenge response of any hardware key, with different parameters. This would allow to use the same intent for a FIDO2 key. We should contact other app creators who would be interested in creating conventions.

**Comment by J-Jamet** [comment]
OK, as ykDroid has not merged the changes yet. I created a new application based on the driver code which will have improvements (virtual Yubikey key with recovery key use in case of loss) and will also integrate the drivers for FIDO2. There is still a lot of work to do and Google has not yet validated the first upload of the beta but it will allow to test in good conditions.
I will probably need help to implement the new features.

**Comment by J-Jamet** [comment]
Here is an overview of unlocking a Yubikey in USB OTG with the new driver application : 

![cb2d23](https://user-images.githubusercontent.com/19224038/184122644-f2f8ce4f-80a4-4ee3-a019-a65d0b880ba5.gif)

**Comment by pmorange** [comment]
Looks real good @J-Jamet 

For me, an important aspect of the implementation would be how often the key would need to be used for quick unlock or that kind of feature.
As I said in a comment in another issue, I'm currently using KeePassDX's competitor and although I have "Quick Unlock" activated, the app forces me to use the Yubikey way too often (and each time enter my very long unlock password too). That's why I'm looking for alternatives.
I don't know if this is really possible, but once unlocked with Yubikey the ideal would be to be able to unlock the DB with fingerprint only until phone is restarted : in that case the Yubikey would be necessary to unlock DB again.
Some will prefer stronger security but I like to be able to unlock DB on my phone without needing to always have the Yubikey near me...

Thanks for the good work.

**Comment by J-Jamet** [comment]
> For me, an important aspect of the implementation would be how often the key would need to be used for quick unlock or that kind of feature.

Pretty simple to answer : There is no quick unlock and the functionality will not be integrated for the reasons that are in the wiki. https://github.com/Kunzisoft/KeePassDX/wiki/Advanced-Unlocking#why-not-quick-unlock

> I don't know if this is really possible, but once unlocked with Yubikey the ideal would be to be able to unlock the DB with fingerprint only until phone is restarted : in that case the Yubikey would be necessary to unlock DB again.

This functionality is under study, for the moment only the password can be associated with a fingerprint. I plan to extend advanced unlock to key files and hardware keys but I can't say without having the technical constraints after testing. Linked to https://github.com/Kunzisoft/KeePassDX/issues/427 https://github.com/Kunzisoft/KeePassDX/issues/1315

> Some will prefer stronger security but I like to be able to unlock DB on my phone without needing to always have the Yubikey near me...

In this case, there is a limited interest in using it. The purpose of a challenge response is to generate from new keys each time a different response understandable by the challenge emitter. In other words, it is meant to certify a physical presence of the user. 

If you want to store the final decryption key in the client system for ease of use, there is no need to have a physical authentication system for the user since you have stored his authentication system locally. It can be stored temporarily, but you have to keep in mind that this goes against the basic principle of the feature.

Another point can be frightening: at each save, it is necessary to reuse the Yubikey to generate a new key. But for me it's normal because otherwise there would be no interest to use this key. It's like every time, people want security AND ease of use but it's always a choice to make.

**Comment by jivarson** [comment]
> Here is an overview of unlocking a database with a Yubikey in USB OTG and the new driver application :
> 
> ![cb2d23](https://user-images.githubusercontent.com/19224038/184122644-f2f8ce4f-80a4-4ee3-a019-a65d0b880ba5.gif) [ ![cb2d23](https://user-images.githubusercontent.com/19224038/184122644-f2f8ce4f-80a4-4ee3-a019-a65d0b880ba5.gif) ](https://user-images.githubusercontent.com/19224038/184122644-f2f8ce4f-80a4-4ee3-a019-a65d0b880ba5.gif) [ ](https://user-images.githubusercontent.com/19224038/184122644-f2f8ce4f-80a4-4ee3-a019-a65d0b880ba5.gif)

Is this NFC compatible as well?

**Comment by rugk** [comment]
Note for NFC compatibility, you may have a look at https://github.com/microg/GmsCore/issues/849, as there as far as I see such a thing is also currently being implemented.

**Comment by J-Jamet** [comment]
> Is this NFC compatible as well?

Yes, I need to do more tests but it should work. Thx @rugk for the link.

**Comment by J-Jamet** [comment]
I just uploaded the beta of version 3.5.0. Those who are subscribed to the beta on the Play Store will be able to use their Yubikey. But Google doesn't want to validate the Key Driver application so it's only downloadable in debug version for the moment.

**Comment by jivarson** [comment]
> I just uploaded the beta of version 3.5.0. Those who are subscribed to the beta on the Play Store will be able to use their Yubikey. But Google doesn't want to validate the Key Driver application so it's only downloadable in debug version for the moment.

It's works, awesome!!
Just created a kdbx with the 3.5.0 and your Key Driver from Play store.
I'll make a copy of my main vault and use this as a daily driver to report if anything breaks.

**Comment by exstntlmsnthrp** [comment]
> I just uploaded the beta of version 3.5.0. Those who are subscribed to the beta on the Play Store will be able to use their Yubikey. But Google doesn't want to validate the Key Driver application so it's only downloadable in debug version for the moment.

Not all of us use the Play Store. When will this be available to the rest of us? And where does one find the Key Driver app? Maybe I looked in the wrong place, but didn't find it. Cheers.

**Comment by J-Jamet** [comment]
https://github.com/Kunzisoft/KeePassDX/releases/tag/2.5.0beta01
https://github.com/Kunzisoft/KeePassDX/issues/1080

**Comment by agates** [comment]
> I just uploaded the beta of version 3.5.0. Those who are subscribed to the beta on the Play Store will be able to use their Yubikey. But Google doesn't want to validate the Key Driver application so it's only downloadable in debug version for the moment.

I've got the Key Driver installed from Google Play.  It seems to either be crashing or not loading correctly when KeePassDX is waiting for the challenge response.

Happy to open an issue in Gitlab, just let me know what information I need or if there's a way to provide you debug logs.

Also, thank you for your work!  I've purchased "KeePro" to support you.

**Comment by macearl** [comment]
For me it also crashes after attaching the yubikey/while waiting for the challenge response.

Not sure if you want additional information here or over on gitlab in the hardware key driver repo.

Android Version: LineageOS for microG 19.1
Device: Xiaomi Pocophone F1
Stacktrace:
```
FATAL EXCEPTION: main
Process: com.kunzisoft.hardware.key, PID: 21629
java.lang.RuntimeException: Error receiving broadcast Intent { act=android.hardware.usb.action.USB_DEVICE_ATTACHED flg=0x11000010 (has extras) } in com.kunzisoft.hardware.key.ConnectionManager@d846016
	at android.app.LoadedApk$ReceiverDispatcher$Args.lambda$getRunnable$0$LoadedApk$ReceiverDispatcher$Args(LoadedApk.java:1689)
	at android.app.LoadedApk$ReceiverDispatcher$Args$$ExternalSyntheticLambda0.run(Unknown Source:2)
	at android.os.Handler.handleCallback(Handler.java:938)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:201)
	at android.os.Looper.loop(Looper.java:288)
	at android.app.ActivityThread.main(ActivityThread.java:7870)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
Caused by: java.lang.IllegalArgumentException: com.kunzisoft.hardware.key: Targeting S+ (version 31 and above) requires that one of FLAG_IMMUTABLE or FLAG_MUTABLE be specified when creating a PendingIntent.
Strongly consider using FLAG_IMMUTABLE, only use FLAG_MUTABLE if some functionality depends on the PendingIntent being mutable, e.g. if it needs to be used with inline replies or bubbles.
	at android.app.PendingIntent.checkFlags(PendingIntent.java:375)
	at android.app.PendingIntent.getBroadcastAsUser(PendingIntent.java:645)
	at android.app.PendingIntent.getBroadcast(PendingIntent.java:632)
	at com.kunzisoft.hardware.key.ConnectionManager.requestPermission(ConnectionManager.java:242)
	at com.kunzisoft.hardware.key.ConnectionManager.onReceive(ConnectionManager.java:182)
	at android.app.LoadedApk$ReceiverDispatcher$Args.lambda$getRunnable$0$LoadedApk$ReceiverDispatcher$Args(LoadedApk.java:1679)
	... 9 more
```

**Comment by J-Jamet** [comment]
Thanks for the logs, I will try to solve the problem quickly in the Driver Key App.

**Comment by nagromc** [comment]
I just installed Key Driver alongside KeePassDX, and KeePro Unlocker, and tried to unlock my password DB. Unfortunately, I cannot open it.

Here is what I’ve done:
- enter my master password
- choose _Yubikey challenge-response_
- tap the _Open_ button
- a _Loading database…_ dialog opened, with the text _Waiting for the challenge response…_
- swipe my Yubikey behind my phone, I hear a chord (that is not the one I can hear usually with Keepass2Android)
- the dialog is still there, and the ring is still spinning

If you require a stacktrace, please tell me how to get it.

**Comment by jivarson** [comment]
One thing that isn't possible either with this Key Driver nor YkDroid is to use Yubikey on a Chromebook with USB-A (without NFC). 
I guess this isn't something easy to fix in Key Driver?

**Comment by J-Jamet** [comment]
@macearl 
It seems to be a problem with pending intent flags, I will update the call in version 0.1.3 to avoid the crash on your device.

@nagromc 
It seems to be a problem with the detection of the driver application during the query in the database opening thread. It happens if the intent request in the thread is not received, in this case you have to press the "Cancel" button and start the database opening again. The goal is to have access to the hardware key detection application in the database opening workflow, which is put on hold until the response is received.

@jivarson 
> I guess this isn't something easy to fix in Key Driver?

I don't have a chomebook, so I can't test. Maybe version 0.1.3 will solve the problem. If not, someone with the hardware and knowledge will have to look into the problem, but theoretically it should work, but there may be weird incompatibilities.

**Comment by super9mega** [comment]
following your gitlab pipelines, installing the debug app, I'm able to get it to not crash like before but now I'm getting this crash after I scan the key
```09-15 15:07:58.985  1956 20940 I ActivityTaskManager: START u0 {cmp=com.kunzisoft.hardware.key/.ChallengeResponseActivity} from uid 10530
09-15 15:07:58.987  1956 20940 W ActivityTaskManager: startActivity called from non-Activity context; forcing Intent.FLAG_ACTIVITY_NEW_TASK for: Intent { cmp=com.kunzisoft.hardware.key/.ChallengeResponseActivity }
09-15 15:07:58.993 16320 16320 D AndroidRuntime: Shutting down VM
09-15 15:07:58.995 16320 16320 E AndroidRuntime: FATAL EXCEPTION: main
09-15 15:07:58.995 16320 16320 E AndroidRuntime: Process: com.kunzisoft.hardware.key, PID: 16320
09-15 15:07:58.995 16320 16320 E AndroidRuntime: java.lang.AssertionError
09-15 15:07:58.995 16320 16320 E AndroidRuntime:        at com.kunzisoft.hardware.key.ConnectionManager.onReceive(ConnectionManager.java:186)
09-15 15:07:58.995 16320 16320 E AndroidRuntime:        at com.kunzisoft.hardware.key.ChallengeResponseActivity.onNewIntent(ChallengeResponseActivity.kt:166)
09-15 15:07:58.995 16320 16320 E AndroidRuntime:        at android.app.Activity.performNewIntent(Activity.java:8306)
09-15 15:07:58.995 16320 16320 E AndroidRuntime:        at android.app.Instrumentation.callActivityOnNewIntent(Instrumentation.java:1490)
09-15 15:07:58.995 16320 16320 E AndroidRuntime:        at android.app.Instrumentation.callActivityOnNewIntent(Instrumentation.java:1503)
09-15 15:07:58.995 16320 16320 E AndroidRuntime:        at android.app.ActivityThread.deliverNewIntents(ActivityThread.java:3853)
09-15 15:07:58.995 16320 16320 E AndroidRuntime:        at android.app.ActivityThread.handleNewIntent(ActivityThread.java:3860)
09-15 15:07:58.995 16320 16320 E AndroidRuntime:        at android.app.servertransaction.NewIntentItem.execute(NewIntentItem.java:56)
09-15 15:07:58.995 16320 16320 E AndroidRuntime:        at android.app.servertransaction.ActivityTransactionItem.execute(ActivityTransactionItem.java:45)
09-15 15:07:58.995 16320 16320 E AndroidRuntime:        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
09-15 15:07:58.995 16320 16320 E AndroidRuntime:        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
09-15 15:07:58.995 16320 16320 E AndroidRuntime:        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2308)
09-15 15:07:58.995 16320 16320 E AndroidRuntime:        at android.os.Handler.dispatchMessage(Handler.java:106)
09-15 15:07:58.995 16320 16320 E AndroidRuntime:        at android.os.Looper.loopOnce(Looper.java:201)
09-15 15:07:58.995 16320 16320 E AndroidRuntime:        at android.os.Looper.loop(Looper.java:288)
09-15 15:07:58.995 16320 16320 E AndroidRuntime:        at android.app.ActivityThread.main(ActivityThread.java:7894)
09-15 15:07:58.995 16320 16320 E AndroidRuntime:        at java.lang.reflect.Method.invoke(Native Method)
09-15 15:07:58.995 16320 16320 E AndroidRuntime:        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:550)
09-15 15:07:58.995 16320 16320 E AndroidRuntime:        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
09-15 15:07:58.996  1311  1311 D vendor.qti.vibrator: QTI Vibrator on for 200 ms with a gain of 0x80
09-15 15:07:58.999  1956  2038 W ActivityTaskManager:   Force finishing activity com.kunzisoft.hardware.key/.ChallengeResponseActivity
09-15 15:07:59.012 16320 16320 I Process : Sending signal. PID: 16320 SIG: 9
09-15 15:07:59.047  1956  4307 I WindowManager: WIN DEATH: Window{df46663 u0 com.kunzisoft.hardware.key/com.kunzisoft.hardware.key.ChallengeResponseActivity}
09-15 15:07:59.047  1956  4307 W InputManager-JNI: Input channel object 'df46663 com.kunzisoft.hardware.key/com.kunzisoft.hardware.key.ChallengeResponseActivity (client)' was disposed without first being removed with the input manager!
09-15 15:07:59.048  1956  2038 I ActivityManager: Process com.kunzisoft.hardware.key (pid 16320) has died: fg  TOP
09-15 15:07:59.048  1078  1078 I Zygote  : Process 16320 exited due to signal 9 (Killed)
09-15 15:07:59.071 12436 12436 E HardwareKeyResponseHelper: Response from challenge error
09-15 15:07:59.095  1956  2276 W WindowManager: Failed to deliver inset state change to w=Window{df46663 u0 com.kunzisoft.hardware.key/com.kunzisoft.hardware.key.ChallengeResponseActivity EXITING}
09-15 15:07:59.095  1956  2276 W WindowManager: android.os.DeadObjectException
09-15 15:07:59.095  1956  2276 W WindowManager:         at android.os.BinderProxy.transactNative(Native Method)
09-15 15:07:59.095  1956  2276 W WindowManager:         at android.os.BinderProxy.transact(BinderProxy.java:584)
09-15 15:07:59.095  1956  2276 W WindowManager:         at android.view.IWindow$Stub$Proxy.insetsControlChanged(IWindow.java:473)
09-15 15:07:59.095  1956  2276 W WindowManager:         at com.android.server.wm.WindowState.notifyInsetsControlChanged(WindowState.java:3994)
09-15 15:07:59.095  1956  2276 W WindowManager:         at com.android.server.wm.InsetsStateController.lambda$notifyPendingInsetsControlChanged$5$com-android-server-wm-InsetsStateController(InsetsStateController.java:348)
09-15 15:07:59.095  1956  2276 W WindowManager:         at com.android.server.wm.InsetsStateController$$ExternalSyntheticLambda4.run(Unknown Source:2)
09-15 15:07:59.095  1956  2276 W WindowManager:         at com.android.server.wm.WindowAnimator.executeAfterPrepareSurfacesRunnables(WindowAnimator.java:345)
09-15 15:07:59.095  1956  2276 W WindowManager:         at com.android.server.wm.WindowAnimator.animate(WindowAnimator.java:226)
09-15 15:07:59.095  1956  2276 W WindowManager:         at com.android.server.wm.WindowAnimator.lambda$new$1$com-android-server-wm-WindowAnimator(WindowAnimator.java:106)
09-15 15:07:59.095  1956  2276 W WindowManager:         at com.android.server.wm.WindowAnimator$$ExternalSyntheticLambda1.doFrame(Unknown Source:2)
09-15 15:07:59.095  1956  2276 W WindowManager:         at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1229)
09-15 15:07:59.095  1956  2276 W WindowManager:         at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1239)
09-15 15:07:59.095  1956  2276 W WindowManager:         at android.view.Choreographer.doCallbacks(Choreographer.java:899)
09-15 15:07:59.095  1956  2276 W WindowManager:         at android.view.Choreographer.doFrame(Choreographer.java:827)
09-15 15:07:59.095  1956  2276 W WindowManager:         at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1214)
09-15 15:07:59.095  1956  2276 W WindowManager:         at android.os.Handler.handleCallback(Handler.java:942)
09-15 15:07:59.095  1956  2276 W WindowManager:         at android.os.Handler.dispatchMessage(Handler.java:99)
09-15 15:07:59.095  1956  2276 W WindowManager:         at android.os.Looper.loopOnce(Looper.java:201)
09-15 15:07:59.095  1956  2276 W WindowManager:         at android.os.Looper.loop(Looper.java:288)
09-15 15:07:59.095  1956  2276 W WindowManager:         at android.os.HandlerThread.run(HandlerThread.java:67)
09-15 15:07:59.095  1956  2276 W WindowManager:         at com.android.server.ServiceThread.run(ServiceThread.java:44)
09-15 15:07:59.112  1956  2276 W WindowManager: Exception thrown during dispatchAppVisibility Window{df46663 u0 com.kunzisoft.hardware.key/com.kunzisoft.hardware.key.ChallengeResponseActivity EXITING}
09-15 15:07:59.112  1956  2276 W WindowManager: android.os.DeadObjectException
09-15 15:07:59.112  1956  2276 W WindowManager:         at android.os.BinderProxy.transactNative(Native Method)
09-15 15:07:59.112  1956  2276 W WindowManager:         at android.os.BinderProxy.transact(BinderProxy.java:584)
09-15 15:07:59.112  1956  2276 W WindowManager:         at android.view.IWindow$Stub$Proxy.dispatchAppVisibility(IWindow.java:536)
09-15 15:07:59.112  1956  2276 W WindowManager:         at com.android.server.wm.WindowState.sendAppVisibilityToClients(WindowState.java:3489)
09-15 15:07:59.112  1956  2276 W WindowManager:         at com.android.server.wm.WindowContainer.sendAppVisibilityToClients(WindowContainer.java:1199)
09-15 15:07:59.112  1956  2276 W WindowManager:         at com.android.server.wm.WindowToken.setClientVisible(WindowToken.java:398)
09-15 15:07:59.112  1956  2276 W WindowManager:         at com.android.server.wm.ActivityRecord.setClientVisible(ActivityRecord.java:6564)
09-15 15:07:59.112  1956  2276 W WindowManager:         at com.android.server.wm.ActivityRecord.onAnimationFinished(ActivityRecord.java:7351)
09-15 15:07:59.112  1956  2276 W WindowManager:         at com.android.server.wm.ActivityRecord.postApplyAnimation(ActivityRecord.java:5263)
09-15 15:07:59.112  1956  2276 W WindowManager:         at com.android.server.wm.ActivityRecord.commitVisibility(ActivityRecord.java:5235)
09-15 15:07:59.112  1956  2276 W WindowManager:         at com.android.server.wm.ActivityRecord.commitVisibility(ActivityRecord.java:5239)
09-15 15:07:59.112  1956  2276 W WindowManager:         at com.android.server.wm.AppTransitionController.handleClosingApps(AppTransitionController.java:1098)
09-15 15:07:59.112  1956  2276 W WindowManager:         at com.android.server.wm.AppTransitionController.handleAppTransitionReady(AppTransitionController.java:286)
09-15 15:07:59.112  1956  2276 W WindowManager:         at com.android.server.wm.RootWindowContainer.checkAppTransitionReady(RootWindowContainer.java:975)
09-15 15:07:59.112  1956  2276 W WindowManager:         at com.android.server.wm.RootWindowContainer.performSurfacePlacementNoTrace(RootWindowContainer.java:842)
09-15 15:07:59.112  1956  2276 W WindowManager:         at com.android.server.wm.RootWindowContainer.performSurfacePlacement(RootWindowContainer.java:784)
09-15 15:07:59.112  1956  2276 W WindowManager:         at com.android.server.wm.WindowSurfacePlacer.performSurfacePlacementLoop(WindowSurfacePlacer.java:177)
09-15 15:07:59.112  1956  2276 W WindowManager:         at com.android.server.wm.WindowSurfacePlacer.performSurfacePlacement(WindowSurfacePlacer.java:126)
09-15 15:07:59.112  1956  2276 W WindowManager:         at com.android.server.wm.WindowSurfacePlacer.performSurfacePlacement(WindowSurfacePlacer.java:115)
09-15 15:07:59.112  1956  2276 W WindowManager:         at com.android.server.wm.WindowSurfacePlacer$Traverser.run(WindowSurfacePlacer.java:57)
09-15 15:07:59.112  1956  2276 W WindowManager:         at android.os.Handler.handleCallback(Handler.java:942)
09-15 15:07:59.112  1956  2276 W WindowManager:         at android.os.Handler.dispatchMessage(Handler.java:99)
09-15 15:07:59.112  1956  2276 W WindowManager:         at android.os.Looper.loopOnce(Looper.java:201)
09-15 15:07:59.112  1956  2276 W WindowManager:         at android.os.Looper.loop(Looper.java:288)
09-15 15:07:59.112  1956  2276 W WindowManager:         at android.os.HandlerThread.run(HandlerThread.java:67)
09-15 15:07:59.112  1956  2276 W WindowManager:         at com.android.server.ServiceThread.run(ServiceThread.java:44)
09-15 15:07:59.148 12436 12463 E com.kunzisoft.keepass.database.element.Database: Unable to load the database
09-15 15:07:59.148 12436 12463 E ActionRunnable: success=false
09-15 15:07:59.148 12436 12463 E ActionRunnable: java.util.concurrent.CancellationException: Unable to get the response from the challenge.
09-15 15:07:59.148 12436 12463 E ActionRunnable:        at com.kunzisoft.keepass.services.DatabaseTaskNotificationService.cancelChallengeResponse(DatabaseTaskNotificationService.kt:264)
09-15 15:07:59.148 12436 12463 E ActionRunnable:        at com.kunzisoft.keepass.services.DatabaseTaskNotificationService.access$cancelChallengeResponse(DatabaseTaskNotificationService.kt:62)
09-15 15:07:59.148 12436 12463 E ActionRunnable:        at com.kunzisoft.keepass.services.DatabaseTaskNotificationService$respondToChallenge$1.invokeSuspend(DatabaseTaskNotificationService.kt:273)
09-15 15:07:59.148 12436 12463 E ActionRunnable:        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
09-15 15:07:59.148 12436 12463 E ActionRunnable:        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
09-15 15:07:59.148 12436 12463 E ActionRunnable:        at android.os.Handler.handleCallback(Handler.java:942)
09-15 15:07:59.148 12436 12463 E ActionRunnable:        at android.os.Handler.dispatchMessage(Handler.java:99)
09-15 15:07:59.148 12436 12463 E ActionRunnable:        at android.os.Looper.loopOnce(Looper.java:201)
09-15 15:07:59.148 12436 12463 E ActionRunnable:        at android.os.Looper.loop(Looper.java:288)
09-15 15:07:59.148 12436 12463 E ActionRunnable:        at android.app.ActivityThread.main(ActivityThread.java:7894)
09-15 15:07:59.148 12436 12463 E ActionRunnable:        at java.lang.reflect.Method.invoke(Native Method)
09-15 15:07:59.148 12436 12463 E ActionRunnable:        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:550)
09-15 15:07:59.148 12436 12463 E ActionRunnable:        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
09-15 15:07:59.148 12436 12463 E ActionRunnable: com.kunzisoft.keepass.database.exception.DatabaseInputException
09-15 15:07:59.148 12436 12463 E ActionRunnable:        at com.kunzisoft.keepass.database.element.Database.loadData(Database.kt:624)
09-15 15:07:59.148 12436 12463 E ActionRunnable:        at com.kunzisoft.keepass.database.action.LoadDatabaseRunnable.onActionRun(LoadDatabaseRunnable.kt:56)
09-15 15:07:59.148 12436 12463 E ActionRunnable:        at com.kunzisoft.keepass.tasks.ActionRunnable.run(ActionRunnable.kt:36)
09-15 15:07:59.148 12436 12463 E ActionRunnable:        at com.kunzisoft.keepass.services.DatabaseTaskNotificationService$executeAction$2$1$asyncResult$1.invokeSuspend(DatabaseTaskNotificationService.kt:587)
09-15 15:07:59.148 12436 12463 E ActionRunnable:        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
09-15 15:07:59.148 12436 12463 E ActionRunnable:        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
09-15 15:07:59.148 12436 12463 E ActionRunnable:        at kotlinx.coroutines.internal.LimitedDispatcher.run(LimitedDispatcher.kt:42)
09-15 15:07:59.148 12436 12463 E ActionRunnable:        at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:95)
09-15 15:07:59.148 12436 12463 E ActionRunnable:        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:570)
09-15 15:07:59.148 12436 12463 E ActionRunnable:        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:749)
09-15 15:07:59.148 12436 12463 E ActionRunnable:        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:677)
09-15 15:07:59.148 12436 12463 E ActionRunnable:        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:664)
09-15 15:07:59.166  1956 20940 D CoreBackPreview: Window{d0c6249 u0 com.kunzisoft.keepass.free/com.kunzisoft.keepass.activities.MainCredentialActivity}: Setting back callback null

```

**Comment by macearl** [comment]
Thanks for the quick response.
With 0.13 I can occasionally finish the challenge response, sometime however it still crashes with another intent related(?) problem:

```
FATAL EXCEPTION: main
Process: com.kunzisoft.hardware.key, PID: 9848
java.lang.RuntimeException: Error receiving broadcast Intent { act=android.yubikey.intent.action.USB_PERMISSION_REQUEST flg=0x10 } in com.kunzisoft.hardware.key.ConnectionManager@768bda5
	at android.app.LoadedApk$ReceiverDispatcher$Args.lambda$getRunnable$0$LoadedApk$ReceiverDispatcher$Args(LoadedApk.java:1689)
	at android.app.LoadedApk$ReceiverDispatcher$Args$$ExternalSyntheticLambda0.run(Unknown Source:2)
	at android.os.Handler.handleCallback(Handler.java:938)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:201)
	at android.os.Looper.loop(Looper.java:288)
	at android.app.ActivityThread.main(ActivityThread.java:7870)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'int android.hardware.usb.UsbDevice.getVendorId()' on a null object reference
	at com.kunzisoft.hardware.yubikey.challenge.UsbYubiKey$Type.lookupDeviceType(UsbYubiKey.java:121)
	at com.kunzisoft.hardware.yubikey.challenge.UsbYubiKey$Type.isDeviceKnown(UsbYubiKey.java:131)
	at com.kunzisoft.hardware.key.ConnectionManager.requestPermission(ConnectionManager.java:240)
	at com.kunzisoft.hardware.key.ConnectionManager.onReceive(ConnectionManager.java:193)
	at android.app.LoadedApk$ReceiverDispatcher$Args.lambda$getRunnable$0$LoadedApk$ReceiverDispatcher$Args(LoadedApk.java:1679)
	... 9 more
```

what I forgot earlier: I also have ykdroid and yubico Authenticator installed. Yubico authenticator also auto starts every time I plug in my yubikey (so far I have not found a way to disable this)

I'll see if I can find another android device to test on without those apps installed, if no one else experiences the same problem.

**Comment by J-Jamet** [comment]
Thank you @super9mega and @macearl , your logs are very useful.
There are assertions to remove and null checks to manage, I wanted to keep the java code for simplicity but it makes more sense to migrate it to Kotlin, I will do that when I have time. The problem is that I don't have a recent device with Android 12 or 13 to test in real conditions so it's a bit hard to debug the driver.

**Comment by nagromc** [comment]
> @nagromc It seems to be a problem with the detection of the driver application during the query in the database opening thread. It happens if the intent request in the thread is not received, in this case you have to press the "Cancel" button and start the database opening again. The goal is to have access to the hardware key detection application in the database opening workflow, which is put on hold until the response is received.

I’ve tried multiple times, but I hear the same ["failing" chord](https://youtu.be/nTRlhElkcPs?t=31) every time.

With KP2A, [ykDroid prompt](https://play-lh.googleusercontent.com/R2IypB03RnMGqzMzxB7ovEv_FL_-jogeMGNp9O8ihQfgAeWC3SGhJzq8XtMGLlwM5_F5=w2560-h1440-rw) is displayed on top of KP2A to ask the user to swipe his Yubikey. When I try to swipe the Yubikey out of this prompt (say, from the desktop, i.e. with no app launched), I hear the ["failing" chord](https://youtu.be/nTRlhElkcPs?t=31).

I noted that if I install the [Pass Pass Easy Card](https://play.google.com/store/apps/details?id=keolis.example.gse.keolislecteurv1) app (an app to check how many dematerialized public transport ticket we have), when I try to unlock my KDBX, I can hear the usual sound when it succeeds to read the Yubikey. I suspect the Pass Pass Easy Card app to read the NFC chip, instead of KeePassDX.

**Comment by nagromc** [comment]
@J-Jamet I’ve already tried to uninstall the Pass Pass Easy Card app, but the failing chord rings. It is like the Key Driver just ignores the Yubikey. Would a stacktrace help? If so, could you give me a link explaining how to get it?

**Comment by J-Jamet** [comment]
I would like to be sure that the problem comes from Key Driver, KeePassDX 4.5.0beta should work with this modified ykDroid which allows the management of the new intent. Can you confirm it?
[yKDroid_2.1.2_mod.zip](https://github.com/Kunzisoft/KeePassDX/files/9582882/yKDroid_2.1.2_mod.zip)

(Logs : https://www.xda-developers.com/how-to-take-logs-android/)

**Comment by nagromc** [comment]
It works with ykDroid v1.2.1_mod.

Here is a capture of KeePassDX with ykDroid 1.2.1_mod and with Key Driver: https://drive.proton.me/urls/JEVJA27TBM#eeWqLkhfiMcE

This stracktrace may help (thanks for the link):

```
09-16 10:19:18.299 29465 29465 E AndroidRuntime: FATAL EXCEPTION: main
09-16 10:19:18.299 29465 29465 E AndroidRuntime: Process: com.kunzisoft.hardware.key, PID: 29465
09-16 10:19:18.299 29465 29465 E AndroidRuntime: java.lang.RuntimeException: Unable to resume activity {com.kunzisoft.hardware.key/com.kunzisoft.hardware.key.ChallengeResponseActivity}: java.lang.IllegalArgumentException: com.kunzisoft.hardware.key: Targeting S+ (version 31 and above) requires that one of FLAG_IMMUTABLE or FLAG_MUTABLE be specified when creating a PendingIntent.
09-16 10:19:18.299 29465 29465 E AndroidRuntime: Strongly consider using FLAG_IMMUTABLE, only use FLAG_MUTABLE if some functionality depends on the PendingIntent being mutable, e.g. if it needs to be used with inline replies or bubbles.
09-16 10:19:18.299 29465 29465 E AndroidRuntime: 	at android.app.ActivityThread.performResumeActivity(ActivityThread.java:4935)
09-16 10:19:18.299 29465 29465 E AndroidRuntime: 	at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:4974)
09-16 10:19:18.299 29465 29465 E AndroidRuntime: 	at android.app.servertransaction.ResumeActivityItem.execute(ResumeActivityItem.java:60)
09-16 10:19:18.299 29465 29465 E AndroidRuntime: 	at android.app.servertransaction.ActivityTransactionItem.execute(ActivityTransactionItem.java:45)
09-16 10:19:18.299 29465 29465 E AndroidRuntime: 	at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:176)
09-16 10:19:18.299 29465 29465 E AndroidRuntime: 	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
09-16 10:19:18.299 29465 29465 E AndroidRuntime: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2345)
09-16 10:19:18.299 29465 29465 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:106)
09-16 10:19:18.299 29465 29465 E AndroidRuntime: 	at android.os.Looper.loopOnce(Looper.java:233)
09-16 10:19:18.299 29465 29465 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:344)
09-16 10:19:18.299 29465 29465 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:8212)
09-16 10:19:18.299 29465 29465 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
09-16 10:19:18.299 29465 29465 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:584)
09-16 10:19:18.299 29465 29465 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1034)
09-16 10:19:18.299 29465 29465 E AndroidRuntime: Caused by: java.lang.IllegalArgumentException: com.kunzisoft.hardware.key: Targeting S+ (version 31 and above) requires that one of FLAG_IMMUTABLE or FLAG_MUTABLE be specified when creating a PendingIntent.
09-16 10:19:18.299 29465 29465 E AndroidRuntime: Strongly consider using FLAG_IMMUTABLE, only use FLAG_MUTABLE if some functionality depends on the PendingIntent being mutable, e.g. if it needs to be used with inline replies or bubbles.
09-16 10:19:18.299 29465 29465 E AndroidRuntime: 	at android.app.PendingIntent.checkFlags(PendingIntent.java:378)
09-16 10:19:18.299 29465 29465 E AndroidRuntime: 	at android.app.PendingIntent.getActivityAsUser(PendingIntent.java:461)
09-16 10:19:18.299 29465 29465 E AndroidRuntime: 	at android.app.PendingIntent.getActivity(PendingIntent.java:447)
09-16 10:19:18.299 29465 29465 E AndroidRuntime: 	at android.app.PendingIntent.getActivity(PendingIntent.java:411)
09-16 10:19:18.299 29465 29465 E AndroidRuntime: 	at com.kunzisoft.hardware.key.ConnectionManager.initNFCConnection(ConnectionManager.java:134)
09-16 10:19:18.299 29465 29465 E AndroidRuntime: 	at com.kunzisoft.hardware.key.ConnectionManager.onActivityResumed(ConnectionManager.java:108)
09-16 10:19:18.299 29465 29465 E AndroidRuntime: 	at android.app.Application.dispatchActivityResumed(Application.java:430)
09-16 10:19:18.299 29465 29465 E AndroidRuntime: 	at android.app.Activity.dispatchActivityResumed(Activity.java:1400)
09-16 10:19:18.299 29465 29465 E AndroidRuntime: 	at android.app.Activity.onResume(Activity.java:1966)
09-16 10:19:18.299 29465 29465 E AndroidRuntime: 	at androidx.fragment.app.FragmentActivity.onResume(FragmentActivity.java:310)
09-16 10:19:18.299 29465 29465 E AndroidRuntime: 	at android.app.Instrumentation.callActivityOnResume(Instrumentation.java:1491)
09-16 10:19:18.299 29465 29465 E AndroidRuntime: 	at android.app.Activity.performResume(Activity.java:8280)
09-16 10:19:18.299 29465 29465 E AndroidRuntime: 	at android.app.ActivityThread.performResumeActivity(ActivityThread.java:4925)
09-16 10:19:18.299 29465 29465 E AndroidRuntime: 	... 13 more
```

**Comment by J-Jamet** [comment]
I released version 0.1.4 of Key Driver, version 3.5.0Beta02 should be validated by Google in beta soon. Normally there should be no more crashes. Don't hesitate to give feedback. ;)

**Comment by super9mega** [comment]
So! Seems to have solved the crashing (not by my computer so I can't get a log) but now it just makes the sound and never moves on towards opening the database. Also, it's not the connect sound, it's the same sound it makes when you do it without any app open. I'll post my logs when I'm next to a computer (pixel 6 pro, Android 13)

**Comment by wandering-tales** [comment]
I still suffer crashes after upgrading both Key Driver 0.1.4 and KeePassDX to 3.5.0beta02.

As soon as I plug my YubiKey 5C, my Yubico Authenticator app is opened. Then if I switch back to the KeyPassDX app, I am prompted to allow the Key Driver to use the YubiKey, I press OK and the app crashes. Here's some screenshot:

![Screenshot_20221016-230152.png](https://user-images.githubusercontent.com/24679086/196058252-650580f9-03cb-48be-a109-97c6efdfe037.png)

![Screenshot_20221016-230230.png](https://user-images.githubusercontent.com/24679086/196058258-404f8eb1-8f80-48c8-91d4-10d07229f5e4.png)

I have also sent a crash report via Android.

**Comment by J-Jamet** [comment]
It's strange, the crash reports are only for the 0.1.3 versions of Key Driver but I don't have any crash report for the 0.1.4 version.

**Comment by wandering-tales** [comment]
I confirm I have version 0.1.4 of the driver installed. I'll repeat the whole process and I'll send again a crash report.

**Comment by wandering-tales** [comment]
I reproduced the crash and sent a report. I wonder whether these reports are sent to Google at first and then forwarded to developers. I am not an Android developer myself.

**Comment by wandering-tales** [comment]
I wonder whether the Yubico Authenticator to conflicts somehow with the Key Driver, as maybe it requires exclusive access to the YubiKey.

**Comment by macearl** [comment]
> I wonder whether the Yubico Authenticator to conflicts somehow with the Key Driver, as maybe it requires exclusive access to the YubiKey.

Haven't had a chance to test the new version yet, but the yubico authenticator works fine in combination with another app, e.g. ykdroid. So it shouldn't be an exclusive access problem. 

The only hassle is that the authenticator auto starts and you have to manually go back to the previous app and allow the access to the yubikey

**Comment by intrnl** [comment]
Would it be possible for the biometric unlock function to not require having the YubiKey inserted? I kinda wanted it so that my YubiKey is only required when trying to open my password database for the first time.

**Comment by Michal78900** [comment]
Key Driver crashes or throws error when trying to use a Yubikey via NFC. However, I was able to open my database when connected via USB using an adapter. Not sure if NFC not working is a bug or is it just me who doesn't know how to hold the key (I'm fairly new YubiKey user).

I also would appreciate an option for biometric unlock function to not require Yubikey challenge-response, unless the user wants to write changes into the database.

**Comment by J-Jamet** [comment]
> Would it be possible for the biometric unlock function to not require having the YubiKey inserted?

In this case, there is no point in using a Yubikey : https://github.com/Kunzisoft/KeePassDX/issues/8#issuecomment-1212875135

> Key Driver crashes or throws error when trying to use a Yubikey via NFC.

https://gitlab.com/kunzisoft/android-hardware-key-driver/-/issues Download the YkDroid mode while waiting for the update. https://github.com/Kunzisoft/KeePassDX/issues/8#issuecomment-1249168078

**Comment by Michal78900** [comment]
> Download the YkDroid mode while waiting for the update. https://github.com/Kunzisoft/KeePassDX/issues/8#issuecomment-1249168078

Thank you very much! Now it works without any issues via both USB and NFC.

**Comment by ZenMasta** [comment]
Key Driver is working great for me. KPDX build 3.5.0Beta02. It's as fast as on desktop. 

Normally using these keys on phone is super tedious and not nearly as fast as totp because it's a 3-step process. 

1) Use your key on www.website "Get Started"
2) choose how to use (BT, NFC, USB)
3) Google Play Services prompt, allow GPS to access my key

**Comment by j6b72** [comment]
Google Pixel 6P: The provided YkDroid-Mod works flawlessly via NFC, while the Key Driver just vibrates as soon as I swipe my key. No interesting stuff from logcat, but I can provide the logs if required.

Either way, thank you so much for your efforts!

**Comment by pmorange** [comment]
Hi, currently not working on a Galaxy S22 Ultra.
I have the latest beta of KeepassDX and the latest version of the Key Driver.
What happens is ... nothing : the screen changes for me to pass my Yubikey. I get it close to my phone, it vibrates, but then nothing happens. The screen of the key driver keeps on waiting. I'm not sure how I can get logs to you.
Also, when I go back the password field gets resetted.

**Comment by macearl** [comment]
I'm also still having problems with the new 0.1.5

I've disabled my yubico authenticator before testing to exclude the automatic app switching as a possible cause.

If I have my yubikey already attached the key driver asks me it touch the button and then tells me there was a problem and I should reconnect my yubikey.

If I reconnect it or if it was not already connected I get the prompt to allow access to the yubikey and as soon as I confirm this the app crashes with the following error:

```
FATAL EXCEPTION: main
Process: com.kunzisoft.hardware.key, PID: 15598
java.lang.RuntimeException: Error receiving broadcast Intent { act=android.yubikey.intent.action.USB_PERMISSION_REQUEST flg=0x10 } in com.kunzisoft.hardware.key.ConnectionManager@dbb852e
	at android.app.LoadedApk$ReceiverDispatcher$Args.lambda$getRunnable$0$LoadedApk$ReceiverDispatcher$Args(LoadedApk.java:1689)
	at android.app.LoadedApk$ReceiverDispatcher$Args$$ExternalSyntheticLambda0.run(Unknown Source:2)
	at android.os.Handler.handleCallback(Handler.java:938)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:201)
	at android.os.Looper.loop(Looper.java:288)
	at android.app.ActivityThread.main(ActivityThread.java:7870)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'int android.hardware.usb.UsbDevice.getVendorId()' on a null object reference
	at com.kunzisoft.hardware.yubikey.challenge.UsbYubiKey$Type.lookupDeviceType(UsbYubiKey.java:121)
	at com.kunzisoft.hardware.yubikey.challenge.UsbYubiKey$Type.isDeviceKnown(UsbYubiKey.java:131)
	at com.kunzisoft.hardware.key.ConnectionManager.requestPermission(ConnectionManager.kt:224)
	at com.kunzisoft.hardware.key.ConnectionManager.onReceive(ConnectionManager.kt:164)
	at android.app.LoadedApk$ReceiverDispatcher$Args.lambda$getRunnable$0$LoadedApk$ReceiverDispatcher$Args(LoadedApk.java:1679)
	... 9 more
```

**Comment by J-Jamet** [comment]
Can you test the latest release 0.1.7, it should solve the crash problems and allow better error handling. It is now possible to manually restart an NFC scan.

**Comment by super9mega** [comment]
Hey! just tested it and everything seems to load, no crashing, but then it tells me that it could not read credentials. not sure what the issue is but I can get a log here in a minuite

**Comment by macearl** [comment]
The new version works fine for me as well, tried unlocking the database a couple of times no problems or crashes so far. 

I assume the key driver will also be added to fdroid once it reaches a stable version?

**Comment by super9mega** [comment]
oops, Messed up, did not notice we were using KeepassXC format rather than a sperate database.xml format, which is what keepass normal uses. Everything seems to be working NFC and USB.

**Comment by J-Jamet** [comment]
> I assume the key driver will also be added to fdroid once it reaches a stable version?

Indeed

**Comment by wandering-tales** [comment]
The new 0.1.7 versione definitely solved alla my crashing issues. Amazing job!

**Comment by moya2162** [comment]
I noticed the following issue when trying to open database using magikeyboard:

1. Use Magikeyboard to launch keepassdx.
2. Select database.
3. Type password. Yubikey challenge-response already selected as option.
4. Select Open.
5. Get popup about entering challenge-response, not the key driver app. Scan yubikey but fails. 
![Screenshot_20221230-135839_KeePassDX_1](https://user-images.githubusercontent.com/120048096/210114510-e1011d28-a53a-42f7-9c2c-fc377ee9237a.png)
7. Re-enter password and select open.
8. Key driver app properly asks for yubikey
9. Database opens. 

Using keepassdx 3.5 beta 01 and key driver 0.1.7.

**Comment by dave0003** [comment]
I installed KeePassDX and Key Driver to try with my Yubikey 4.  It is working, though I noticed it works better if I wait until Key Driver prompts for the yubikey, to plug the yubikey in.  Then after plugging it in, it asks for permissions, and it works.  If I try to open the database with the yubikey already plugged in, it fails to open the first time, then asks for permissions, and works the second time.

KeePassDX 3.5.0 Beta01 from F-Droid and Key Driver 0.1.7 from Play Store.

**Comment by J-Jamet** [comment]
Thx @moya2162 & @dave0003 for your feedback, it's very helpful in identifying and solving problems. I'll add these bugs to the kanban so I can deal with them. https://github.com/Kunzisoft/KeePassDX/projects/43

**Comment by leaumar** [comment]
Just wanna add my (positive) experience giving kpdx beta and keydriver a try. I got a yubikey just this week, to try the experience myself and figure out how well I can use it for what things. Compared to webauthn under windows and the different official yubi management apps, making a yubikey-locked db (usb+touch) under keepassxc and then opening it on my phone with nfc has worked without a single surprise or error. Very nice, definitely helps make me feel like I could open up to relying on these gadgets soon without worrying much about support issues. Thank you for the great work, buying keepro as we speak 👍

**Comment by J-Jamet** [comment]
@moya2162 The problem is normally fixed in version 2.5.0beta03, there is a better management of asynchronous events.
There is also database merge management using Yubikey as well as autofill management. I didn't see any bug with this new build. If it's OK, I'll switch to stable in a week.
@dave0003 Your issue is directly related to the driver app which is rewritten in a more modern way (The problem is decoupled from the KeePassDX app)
@leaumar Thank you, it is really appreciated.

**Comment by moya2162** [comment]
@J-Jamet So i installed 3.5b3 and tried to unlock my database and it appears the prompting of the key driver works every time, but scanning (nfc) of my yubikey fails every time. I scan (nfc) my yubikey but the key driver doesn't see it. I get a prompt to open a web link to yubikey's key verification page (standard). 

Using an OTG cable the yubikey works, but not ideal. 

I did a fresh install, wiped app data and cache, uninstalled 3.5b1 (fdroid) and installed 3.5b3 (git).

**Comment by J-Jamet** [comment]
The problem comes rather from the driver / NFC reception. I did not touch this part and it works on 3 of my NFC phones.

**Comment by moya2162** [comment]
False alarm! I have a tasker profile that automatically turns on NFC when keepassdx and key driver are open and turns it off when closed. When i installed the new app it wasn't picked up by tasker so NFC was off when keepassdx was launched. NFC only turned on when key driver was open but at that point it wouldn't work. NFC needs to be on when in keepassdx. If not it fails. 

All worked on my end now!

**Comment by J-Jamet** [comment]
Stable release 3.5.0 done, the driver remains in beta version until finalization but remains functional.
Enjoy

**Comment by 52fitz** [comment]
Nice. Many thanks.

**Comment by nimser** [comment]
After clicking "OPEN" I get the prompt to swipe nfc device, then I'm asked to select a default app. I see only web browser, not the keyDriver app. What am I missing?

I installed 0.1.7

**Comment by J-Jamet** [comment]
Duplicate https://github.com/Kunzisoft/KeePassDX/issues/1504
